## 課題1

### 関数 $f(x) = x^3 + 2x^2 - 5x + 6$ がある

1. f(x)を -5≦x≦3 で図示せよ。
2. 初期値 $x_0=-3,1$ の二通りにおいて、ニュートン法を用いて $f(x)=0$ となるxを求めよ。


In [1]:
# 課題1-1
import matplotlib.pyplot as plt
import numpy as np

def f(x):
    return x**3 + 2*x**2 - 5*x + 6

# -5≦x≦3
x = np.arange(-5, 3.1, 0.1)

# x**3 + 2*x**2 - 5*x + 6
y = f(x)

plt.plot(x, y)
plt.xlabel('x')
plt.ylabel('f(x)')
plt.xlim(-6, 6)
plt.ylim(-50, 50)
plt.show()

<Figure size 640x480 with 1 Axes>

In [2]:
# 課題1-2
import sympy as sym

def newton(x0, dif):
    return x0 - (f(x0)/dif.subs(x, x0))

def calc_newton(x0, dif):
    # 不敵な初期値のためのカウンター
    count = 0

    while True:
        x1 = newton(x0, dif)
        if abs(f(x1)) < 10e-6 or count > 10:
            break
        print(float(x1))
        print(float(f(x1)))
        x0 = x1
        count += 1
    return x1, f(x1)

# 初期値
x0_list = [-3, 1]

for x0 in x0_list:
    # x
    x = sym.symbols('x')
    # 微分値
    dif = sym.diff(f(x))
    # ニュートン法
    x, fx = calc_newton(x0, dif)
    print("x = {} のニュートン法を用いた f(x) = 0 は\nf(x) = {}\n x = {}\nとなる\n".format(x0, float(fx), float(x)))
    

    
    


-4.2
-11.808
-3.820565552699229
-1.4714599308448448
-3.757971282062912
-0.0368260846507598
-3.7563224877074934
-2.520686419640285e-05
x = -3 のニュートン法を用いた f(x) = 0 は
f(x) = -1.1838062181189844e-11
 x = -3.7563213575872454
となる

-1.0
12.0
1.0
4.0
-1.0
12.0
1.0
4.0
-1.0
12.0
1.0
4.0
-1.0
12.0
1.0
4.0
-1.0
12.0
1.0
4.0
-1.0
12.0
x = 1 のニュートン法を用いた f(x) = 0 は
f(x) = 4.0
 x = 1.0
となる



### 結果

以上より、  
$x = -3.7563213575872454$ であるため、初期値 $x_0 = 3$ は適切である。  
$x = 1.0$ であるため、初期値 $x_0 = 1$ は不適切である。